src/libstore/derived-path.cc

```diff
@@ -123,10 +125,15 @@ RealisedPath::Set BuiltPath::toRealisedPaths(Store & store) const
                 for (auto& [outputName, outputPath] : p.outputs) {
                     if (settings.isExperimentalFeatureEnabled(
                                 Xp::CaDerivations)) {
+                        auto drvOutput = get(drvHashes, outputName);
+                        if (!drvOutput)
+                            throw Error(
+                                "the derivation '%s' has unrealised output '%s' (derived-path.cc/toRealisedPaths)",
```

Comment:
IMHO we shouldn't have references to source code in error messages (like `derived-path.cc/toRealisedPaths`) because they're meaningless to users.

If this is a "can't happen" logical error, then it's better to use `assert` (since that's much easier to debug using gdb).

Reply:
I would, but the problem is that exactly that scenario happened which is what prompted me to do this PR... (I don't think an `assert` would print out enough context) ... and I didn't want to fudge the error message to make it unique so it can be associated with exactly one position in the source code where the error stems from...
An alternative would be that `Error` should get the ability to gather a full backtrace upon request (e.g. via an environment variable), and print it.

Reply:
Asserts are easier to debug (for me at least) because if you run in gdb, it stops exactly at the point where the assertion happens. Or you can load the coredump into gdb and get the same. Whereas with exceptions, you have to use `catch throw`, but there are a lot of other exceptions being thrown...

Reply:
follow-up PR: https://github.com/NixOS/nix/pull/6484 (males the error message more usable and fixes `getOr`)

Reply:
btw. didn't change it to `assert` because originally it also wasn't an assert.